### ISSUE TITLE
[Breaking change]: Ambiguous overload resolution

### ISSUE THREAD

**Issue by xtqqczze** [dotnet/docs#44102]
### Description

* https://github.com/dotnet/aspnetcore/pull/55010#issuecomment-2059637820
  * https://csharp.godbolt.org/z/K31fxvPs7

### Version

.NET 9

### Previous behavior

no error

### New behavior

```log
(13,23): error CS0121: The call is ambiguous between the following methods or properties: 'Program.Join(string, params string[])' and 'Program.Join(string, params ReadOnlySpan<string>)'
```

### Type of breaking change

- [ ] **Binary incompatible**: Existing binaries might encounter a breaking change in behavior, such as failure to load or execute, and if so, require recompilation.
- [x] **Source incompatible**: When recompiled using the new SDK or component or to target the new runtime, existing source code might require source changes to compile successfully.
- [ ] **Behavioral change**: Existing binaries might behave differently at run time.

### Reason for change

* https://github.com/dotnet/runtime/pull/101499
* https://github.com/dotnet/runtime/pull/101457
* https://github.com/dotnet/runtime/issues/101261

### Recommended action

* https://github.com/dotnet/aspnetcore/pull/55780/commits/7e353a334d5c1af3bb6f190f3e877a13a985880b

### Feature area

C#

---
[Associated WorkItem - 360652](https://dev.azure.com/msft-skilling/Content/_workitems/edit/360652)

**Comment by CamSoper** [comment]
Hey @xtqqczze -- I'm having to connect a lot of dots here. 

I think the gist is that a new overload was added to `string.Join` -- `string.Join(string, params ReadOnlySpan<string>)` -- that resolves ambiguously with `string.Join(string, params string[])`. 

And I think the suggested resolution is to explicitly cast the parameters.

Is that correct?

And what's the tl;dr reason for the change? I saw the issue but that's a lot of conversation to get caught up on.

**Comment by jozkee** [comment]
The tl;dr for the change can be found in https://github.com/dotnet/runtime/issues/77873.

>APIs where we currently have a params T[]-based overload, there’s a benefit to us adding a params ReadOnlySpan<T> overload: it’ll be preferred by the compiler, and the compiler will aim to stackalloc the span rather than using an array.

The breaking change described here is caused by an unfortunate and rare case. StringValues has implicit operators for `string` and `string[]` which cause conflicts with the `params Span<T>` lang feature.  _Note that this is not exclusive to SearchValues but to any type that declares such implicit operators._ SearchValues happens to be one type, part of ASP.NET.

The affected APIs were described in https://github.com/dotnet/runtime/issues/101261:
```cs
System.String.Concat(params ReadOnlySpan<string?> values)
System.String.Join(char separator, params ReadOnlySpan<string?> value)
System.String.Join(string? separator, params ReadOnlySpan<string?> value)
System.IO.Path.Combine(params ReadOnlySpan<string> paths)
System.IO.Path.Join(params ReadOnlySpan<string?> paths)
System.Text.StringBuilder.AppendJoin(string? separator, params ReadOnlySpan<string?> values)
System.Text.StringBuilder.AppendJoin(char separator, params ReadOnlySpan<string?> values)
```

**Comment by jozkee** [comment]
Yes, to explicitly cast to the most appropriate type. In https://github.com/dotnet/aspnetcore/commit/7e353a334d5c1af3bb6f190f3e877a13a985880b, StringValues was previously resolving to `Join(String? separator, params String?[] value)`, the cast maintains the intended behavior.
